components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AuthenticationService.java

```diff
@@ -0,0 +1,254 @@
+
+        return new AuthServiceRequestWrapper(request, parameters);
+    }
+
+    private AuthServiceResponseWrapper getWrappedResponse(HttpServletResponse response) {
+
+        return new AuthServiceResponseWrapper(response);
+    }
+
+    private AuthServiceResponse processCommonAuthResponse(AuthServiceRequestWrapper request,
+                                                          AuthServiceResponseWrapper response)
+            throws AuthServiceException {
+
+        AuthServiceResponse authServiceResponse = new AuthServiceResponse();
+
+        if (isAuthFlowSuccessful(request)) {
+            handleSuccessAuthResponse(request, response, authServiceResponse);
+        } else if (isAuthFlowFailed(request, response)) {
+            handleFailedAuthResponse(request, response, authServiceResponse);
+        } else if (isAuthFlowIncomplete(request)) {
+            handleIntermediateAuthResponse(request, response, authServiceResponse);
+        } else {
+            throw new AuthServiceException("Unknown authentication flow status: " + request.getAuthFlowStatus());
+        }
+
+        return authServiceResponse;
+    }
+
+    private void handleIntermediateAuthResponse(AuthServiceRequestWrapper request, AuthServiceResponseWrapper response,
+                                                AuthServiceResponse authServiceResponse) throws AuthServiceException {
+
+        authServiceResponse.setSessionDataKey(response.getSessionDataKey());
+        authServiceResponse.setFlowStatus(AuthServiceConstants.FlowStatus.INCOMPLETE);
+        AuthServiceResponseData responseData = new AuthServiceResponseData();
+        boolean isMultiOptionsResponse = request.isMultiOptionsResponse();
+
+        List<AuthenticatorData> authenticatorDataList;
+        if (isMultiOptionsResponse) {
+            responseData.setAuthenticatorSelectionRequired(true);
+            authenticatorDataList = getAuthenticatorBasicData(response.getAuthenticators());
+        } else {
+            authenticatorDataList = request.getAuthInitiationData();
+        }
+        responseData.setAuthenticatorOptions(authenticatorDataList);
+        authServiceResponse.setData(responseData);
+    }
+
+    private void handleSuccessAuthResponse(AuthServiceRequestWrapper request, AuthServiceResponseWrapper response,
+                                           AuthServiceResponse authServiceResponse) throws AuthServiceException {
+
+        authServiceResponse.setSessionDataKey(getFlowCompletionSessionDataKey(request, response));
+        authServiceResponse.setFlowStatus(AuthServiceConstants.FlowStatus.SUCCESS_COMPLETED);
+    }
+
+    private void handleFailedAuthResponse(AuthServiceRequestWrapper request, AuthServiceResponseWrapper response,
+                                          AuthServiceResponse authServiceResponse) throws AuthServiceException {
+
+        // TODO: Improve error handling. Different authenticator seems to pass errors in slightly different ways.
```

Comment:
Error handling needs to be further improved as there seems to be several ways errors are communicated. 
This will be improved in a next PR.